### PR TITLE
Adding a delimiter for opening a scope only temporarily

### LINKED ISSUE #11486: Binding arguments to an interpretation scope works bad
#### Description of the problem

```coq
Declare Scope sc.
Delimit Scope sc with S.
Definition f (_ _ : bool) : bool := false.
Definition g (x : bool) := false.
Notation "1" := false : sc.
Notation "a + b" := (f a b) : sc.

Check (g (1 + 1)%S).
```
The answer is

```coq
g (1 + 1)%S
```

That's ok. But I want to make this `%S` disappear using **Arguments**

```coq
Arguments g x%S.
```
I test again

```coq
Check (g (1 + 1)%S).
```

But the answer is now

```coq
g (1%S + 1%S)
```

#### Coq Version

The Coq Proof Assistant, version 8.12+alpha (January 2020)
compiled on Jan 30 2020 17:11:38 with OCaml 4.07.2+dev0-2018-10-04

### LINKED ISSUE #12157: There should be a way to specify both sorts of scope stacks for both notations and Arguments
#### Description of the problem

Consider:
```coq
Declare Scope foo_scope.
Delimit Scope foo_scope with foo.
Notation "'FOO'" := tt : foo_scope.
Arguments id {_} _%foo.
Notation "'id2' x" := x%foo (at level 10, only parsing).
Check id FOO. (* id FOO : unit *)
Fail Check id (fun x : unit => FOO). (* Error: Unknown interpretation for notation "FOO". *)
Check id2 FOO. (* FOO%foo : unit *)
Check id2 (fun x : unit => FOO). (* fun _ : unit => FOO%foo : unit -> unit *)
```
As far as I know, there's no way to get the `Arguments`-sort of behavior for `Notation`, and no way to get the `Notation`-sort of behavior for `Arguments`.

cc @herbelin , who I believe had a plan for implementing this?

See also #3990 (where I reported my initial confusion about this)

#### Coq Version

8.11.1

### LINKED ISSUE #14305: Binding arguments to a scope vs local scope
As mentioned in the [documentation](https://coq.inria.fr/refman/language/extensions/arguments-command.html#binding-arguments-to-a-scope), when binding arguments to a scope:

> The effect of the scope is limited to the argument itself. It does not propagate to subterms

However the behavior happens to be different when using a local scope:
```coq
Parameter g : bool -> bool.
Check (g (true || false)).

Parameter h : option bool -> bool.
Fail Check (h (Some (true || false))).
Check (h (Some (true || false))%bool).
Arguments h _%bool.
Fail Check (h (Some (true || false))).
```
Is there a natural reason for that? is there a way to simulate the local-scope behavior at the arguments level?

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** feature

Fixes #11486 
Fixes #12157 
Fixes #14305

Reopening of #9123.

This adds a syntax `term%_scope` to tell to activate scope `scope` only for the immediate subterm. This allows for instance to define notations which activate `type_scope` only temporarily as in:
```
Notation "{ x  |  P }" := (sig (fun x => P%_type)) : type_scope.
Check fun P Q => { x : nat & P x * Q x }.
```

Note: We may want this to apply also to `Arguments`. Unfortunately, in `Arguments`, `%` has already the meaning of this new `%_` so we add `%_` as a synonym of `%` there and deprecate `%`. In a few versions we will make `%` in `Arguments` error and in the following version give it the same semantics as in terms.

- [x] Added / updated test-suite

- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in CHANGES.md.
- [x] updated `make doc_gram_rsts`

#### Overlays to be merged before the PR

- [x] https://gitlab.mpi-sws.org/iris/iris/-/merge_requests/1015
- [x] https://github.com/mit-pdos/perennial/pull/55

#### Overlays to be merged in sync with the PR

- [x] https://github.com/ejgallego/coq-serapi/pull/366
- [x] https://github.com/coq-tactician/coq-tactician/pull/68
- [x] https://github.com/coq-community/vscoq/pull/681
- [x] https://github.com/QuickChick/QuickChick/pull/342

### CODE REVIEW

**Comment by proux01** [comment]
@herbelin is the above text of the PR still current? it doesn't seem to reflect the state of the discussion in #9123

**Comment by proux01** [comment]
@ejgallego do you still want to assign this one? Otherwise, I'll assign to myself.

**Comment by herbelin** [comment]
@proux01 : title was wrong. I fixed it. The PR needs maturation. Without time to work currently on it, I moved it to draft.

**Comment by proux01** [comment]
@coqbot run full ci

File: plugins/ltac2/tac2quote.ml [hunk 1]
```diff
@@ -106,7 +106,7 @@ let quote_constr ?delimiters c =
   let loc = Constrexpr_ops.constr_loc c in
   Option.cata
     (List.fold_left (fun c d ->
-          CAst.make ?loc @@ Constrexpr.CDelimiters(Id.to_string d, c))
+          CAst.make ?loc @@ Constrexpr.(CDelimiters(DelimUnboundedScope, Id.to_string d, c))) (* ?? *)
```
Comment (SkySkimmer) [inline plugins/ltac2/tac2quote.ml:109]: Not much choice if we want to have backwards compat
Maybe someday we will have syntax to pick if the scopes are deep

**Comment by SkySkimmer** [comment]
Please link overlays in the PR's opening comment.

**Comment by proux01** [comment]
Sure, I first need to open them.

**Review by proux01 — APPROVED** [review]

**Comment by proux01** [comment]
Let's rerun CI just to be on the safe side and finally merge.

**Comment by Blaisorblade** [comment]
How did this change flip the consensus of "shallow should be the default and ideally have the shorter notation" from https://github.com/coq/coq/pull/9123#issuecomment-877804345 ?

**Comment by Blaisorblade** [comment]
On reflection, I'm not sure how often the deep semantics is necessary in term, if `Bind Scope` is used correctly.

OTOH the shallow semantics seems most appropriate for `Arguments`, since I don't see a compelling example for a deep `Arguments` scope (https://github.com/coq/coq/issues/11486 asks for one but only with an artificial example).

**Comment by proux01** [comment]
@Blaisorblade is it ok to merge this as it is or do you want it to be further discussed?

**Comment by Blaisorblade** [comment]
Thanks for clarifying, I understand the concern.

**Comment by proux01** [comment]
Thanks, I'll merge next week if there are no other comments.

**Comment by proux01** [comment]
@coqbot merge now

**Comment by proux01** [comment]
And another almost five years old PR merged

**Pull request merged**